### PR TITLE
Improve the setup-{macos,ubuntu}.sh scripts to not pollute /usr/local.

### PR DESCRIPTION
**Pull request by damianr99**
The diff changes the setup scripts to install the non-package-manager
dependencies into a local `./deps` directory rather than installing
them globally into `/usr/local/` as the root user.

On MacOS, we now support both `brew` and `macports` package managers.

Note: This change doesn't include the changes the `cmake` to use the
local dependencies, someone who understands cmake will need to do that
part :)

### CODE REVIEW

**Comment by damianr99** [comment]
@majetideepak I've updated the commit to make the needed changes to `CMakeLists.txt` as well along the lines you suggested.  I had to also had to update the include directories, and work around a link problem on Big Sur.   `make release` and `make debug` now work for me.

**Comment by majetideepak** [comment]
Looks like you have to rebase with the main branch. There are conflicts.

**Comment by majetideepak** [comment]
and run `make format-fix` to fix the format-check failure.

**Comment by damianr99** [comment]
@majetideepak done.  CircleCI still fails on macos-build, but it looks to be for reasons unrelated to this change (brew is failing to install a package because one of its dependencies has an expired SSL certificate and curl is failing).

**Comment by majetideepak** [comment]
@damianr99 Looks like other PR macos-builds are passing. Can you take a look?

**Comment by damianr99** [comment]
> @damianr99 Looks like other PR macos-builds are passing. Can you take a look?

From what I can tell, the reason the other PR macos-builds as passing are because they are getting a cache hit on the "restore dependency cache" stage of the pipeline, so the subsequent "install dependencies" becomes a no-op and Brew is not installed from scratch.  E.g. here's a recent working example: https://app.circleci.com/pipelines/github/facebookincubator/velox/1292/workflows/44a590c5-62d2-4a1a-b989-bb3cc8e8d64b/jobs/5199 compared to the that's failing for this PR: https://app.circleci.com/pipelines/github/facebookincubator/velox/1279/workflows/e303cd70-42f3-4c62-9e5e-f9fb47aaae00/jobs/5216

They have different cache keys for "velox-circleci-macos-deps-p-<deleted>", I'm guessing this key is based in part on a hash of the CMakeLists / setup scripts.  This suggests any change at the moment that alters these files will also hit the same problem and fail.

The problem is the initial install of ninja via brew: 
```
curl: (60) SSL certificate problem: certificate has expired
More details here: https://curl.haxx.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
Error: ninja: Failed to download resource "pkg-config"
Download failed: https://pkgconfig.freedesktop.org/releases/pkg-config-0.29.2.tar.gz
```

What's strange here is that I can download the file via curl directly and the certificate seems fine:
```
% curl -v 'https://pkgconfig.freedesktop.org/releases/pkg-config-0.29.2.tar.gz'
*   Trying 131.252.210.176...
* TCP_NODELAY set
* Connected to pkgconfig.freedesktop.org (131.252.210.176) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: CN=pkg-config.freedesktop.org
*  start date: Sep  9 01:35:07 2021 GMT
*  expire date: Dec  8 01:35:06 2021 GMT
*  subjectAltName: host "pkgconfig.freedesktop.org" matched cert's "pkgconfig.freedesktop.org"
*  issuer: C=US; O=Let's Encrypt; CN=R3
*  SSL certificate verify ok.
> GET /releases/pkg-config-0.29.2.tar.gz HTTP/1.1
> Host: pkgconfig.freedesktop.org
> User-Agent: curl/7.64.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Fri, 01 Oct 2021 16:47:46 GMT
< Server: Apache/2.4.38 (Debian)
```

**Comment by wenleix** [comment]
This is great!!! ❤️

**Comment by damianr99** [comment]
Looking at the https://curl.se/docs/sslcerts.html page the curl fail reference more closely, it seems this error might not be because the cert on the origin server we're downloading from that's the problem, but more that the bundled Certificate Authorities on the OS image we are using are now out of date and the CA signing certificate has expired.  Maybe we need a step in the flow here that first updates the OS trusted cert authorities before trying to run the setup script.

**Comment by damianr99** [comment]
Update, I ssh'ed into the CircleCI VM that was failing, and can confirm that it's an OS issue with an expired trust store.

On this machine, curl fails:
```
static:~ distiller$ curl -v 'https://pkgconfig.freedesktop.org/releases/pkg-config-0.29.2.tar.gz'
*   Trying 131.252.210.176...
* TCP_NODELAY set
* Connected to pkgconfig.freedesktop.org (131.252.210.176) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (OUT), TLS alert, certificate expired (557):
* SSL certificate problem: certificate has expired
* Closing connection 0
curl: (60) SSL certificate problem: certificate has expired
More details here: https://curl.haxx.se/docs/sslcerts.html
```

Hunting online, I found this this which looks to be the problem we are hitting:
https://security.stackexchange.com/questions/232445/https-connection-to-specific-sites-fail-with-curl-on-macos

I tried one of workarounds suggested (setting an env var of `CURL_SSL_BACKEND=secure-transport`) , and this does fix the problem.

**Comment by majetideepak** [comment]
@damianr99 thanks very much for the investigation.
@pedroerp @funrollloops looks like we have an imminent issue with macos-build. Any thoughts on the workaround suggested above? Thanks.

**Comment by pedroerp** [comment]
I'll defer the question to @funrollloops, as he's the specialist in this area.

File: CMakeLists.txt [hunk 1]
```diff
@@ -17,6 +17,18 @@ set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/CMake" ${CMAKE_MODULE_PATH})
...
 
+set(DEPENDENCY_INSTALL_PREFIX "deps")
+list(APPEND CMAKE_PREFIX_PATH ${DEPENDENCY_INSTALL_PREFIX})
+include_directories(SYSTEM "${DEPENDENCY_INSTALL_PREFIX}/include")
+
+if(${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
+  # Fix link errors on macOS BigSur, without this link fails unable to find
+  # -lSystem.  Based on same workaround found here:
+  # https://github.com/ponylang/ponyc/pull/3686/commits/03cd3f02ffabbc0691382502e4f0579ade52e47c
+  add_link_options(
```
Comment (funrollloops) [inline CMakeLists.txt:28]: Does this problem occur without this change?

File: CMakeLists.txt [hunk 2]
```diff
@@ -17,6 +17,18 @@ set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/CMake" ${CMAKE_MODULE_PATH})
 # set the project name
 project(velox)
 
+set(DEPENDENCY_INSTALL_PREFIX "deps")
+list(APPEND CMAKE_PREFIX_PATH ${DEPENDENCY_INSTALL_PREFIX})
```
Comment (funrollloops) [inline CMakeLists.txt:21]: CMake already has a variable to specify where to find externally provided dependencies: `CMAKE_PREFIX_PATH`. This should be user specified IMO, except for paths that are managed by the CMake script (e.g. through FetchContent or ExternalProject_Add).

If you really want to provide a default for this, add the default to the Makefile instead.

File: scripts/setup-macos.sh [hunk 3]
```diff
@@ -32,7 +32,11 @@ FB_OS_VERSION=v2021.05.10.00
 NPROC=$(sysctl -n hw.physicalcpu)
 COMPILER_FLAGS="-mavx2 -mfma -mavx -mf16c -masm=intel -mlzcnt"
 DEPENDENCY_DIR=${DEPENDENCY_DIR:-$(pwd)}
-MACOS_DEPS="ninja cmake ccache protobuf icu4c boost gflags glog libevent lz4 lzo snappy xz zstd"
+
+MACOS_DEPS_BREW="ninja cmake ccache protobuf icu4c boost gflags glog libevent lz4 lzo snappy xz zstd"
+MACOS_DEPS_MACPORTS="ninja cmake ccache protobuf3-cpp icu boost gflags google-glog libevent lz4 lzo snappy xz zstd openssl"
```
Comment (funrollloops) [inline scripts/setup-macos.sh:37]: While I welcome support for anything other than `brew`, the `brew` path is tested in the CI and is therefore unlikely to break. The macports path isn't; will you sign up to maintain it?

File: scripts/setup-macos.sh [hunk 4]
```diff
@@ -86,23 +90,33 @@ function cmake_install {
...
   do
-    brew install --formula $pkg && echo "Installation of $pkg is successful" || brew upgrade --formula $pkg
+    sudo port install $pkg
   done
+}
+
+function install_build_prerequisites {
+  if hash brew 2>/dev/null; then
+    install_build_prerequisites_brew
+  else
```
Comment (funrollloops) [inline scripts/setup-macos.sh:117]: Check for `port`; if neither `brew` or `port` is available print an error message asking the user to install one.

File: scripts/setup-macos.sh [hunk 5]
```diff
@@ -117,14 +131,22 @@ function install_fmt {
   cmake_install -DFMT_TEST=OFF
 }
 
+function openssl_dir {
+  if hash brew 2>/dev/null; then
+    brew --prefix openssl
+  else
+    echo /opt/local/lib/openssl
```
Comment (funrollloops) [inline scripts/setup-macos.sh:138]: It's funny to hardcode this path in the same PR that changes the default install locations to something other than /usr/local. Is there an alternative way to handle this, e.g. can we add a find script? Why does OpenSSL require special handling anyway?

File: scripts/setup-macos.sh [hunk 6]
```diff
@@ -86,23 +90,33 @@ function cmake_install {
   mkdir -p "${BINARY_DIR}"
   cmake -Wno-dev -B"${BINARY_DIR}" \
     -GNinja \
+    -DCMAKE_INSTALL_PREFIX="$INSTALL_DEPS_PREFIX" \
     -DCMAKE_CXX_STANDARD=17 \
-    "${INSTALL_PREFIX+-DCMAKE_PREFIX_PATH=}${INSTALL_PREFIX-}" \
```
Comment (funrollloops) [inline scripts/setup-macos.sh:90]: This script already used the `INSTALL_PREFIX` variable to allow the user to specify where libraries could be installed. They would then pass the same path to `CMAKE_PREFIX_PATH` to use their custom location.

Maybe we can address this with better documentation? IMO hard-coded absolute prefix paths in CMake are a bit of a code smell.

**Comment by funrollloops** [comment]
I generally agree with not polluting /usr/local -- in fact I already install the deps to a separate directory. This is how you can do that today:

```
INSTALL_PREFIX=$(pwd)/_build/deps scripts/setup-macos.sh
cmake ... -DCMAKE_PREFIX_PATH=$(pwd)/_build/deps
```

We currently default to installing to `/usr/local` because it provides a simpler first-time dev experience. However, I agree it's a bad default. Change the default INSTALL_PREFIX in the script and pass the appropriate path in the Makefile, but leave the default path out of the CMake script. That'll leave the CMake script and /usr/local unpolluted but hopefully still provide a smooth setup experience.

setup-macos.sh:
```
# If the user did not specify INSTALL_PREFIX, default to the _build directory.
INSTALL_PREFIX=${INSTALL_PREFIX-${BASH_SOURCE[0]}/../_build/prefix
```

Makefile:
```
... -DCMAKE_PREFIX_PATH=${BUILD_BASE_DIR}/prefix
```

Separately -- it would be nice to consolidate the mac, ubuntu, and fedora scripts by automatically detecting the package manager. There's a lot of unnecessary duplication right now IMO.

**Review by funrollloops — CHANGES_REQUESTED** [review]

**Comment by funrollloops** [comment]
> @damianr99 thanks very much for the investigation. @pedroerp @funrollloops looks like we have an imminent issue with macos-build. Any thoughts on the workaround suggested above? Thanks.

Should we just update to a newer macos image? We're currently on XCode 11.7.0, but the newest image is 13.0:
https://circleci.com/docs/2.0/testing-ios/#supported-xcode-versions

@pedroerp Do you remember why we picked this version?

File: CMakeLists.txt (hunk 1 above)
Comment (damianr99) [reply CMakeLists.txt:28]: Yes, on my FB laptop I can't link without this fix
BigSur, 11.6
Darwin mbp 20.6.0 Darwin Kernel Version 20.6.0: Mon Aug 30 06:12:21 PDT 2021; root:xnu-7195.141.6~3/RELEASE_X86_64 x86_64

File: scripts/setup-macos.sh (hunk 3 above)
Comment (damianr99) [reply scripts/setup-macos.sh:37]: I think you just motivated me to install `brew` :)

File: scripts/setup-macos.sh (hunk 5 above)
Comment (damianr99) [reply scripts/setup-macos.sh:138]: This path was hardcoded in the script before my diff, so I didn't change it.  I've no idea why it was done this way.

File: scripts/setup-macos.sh (hunk 3 above)
Comment (funrollloops) [reply scripts/setup-macos.sh:37]: I too both dislike and use brew. Welcome to the club!

File: scripts/setup-macos.sh (hunk 5 above)
Comment (funrollloops) [reply scripts/setup-macos.sh:138]: Ah, I see that now below.

**Comment by mbasmanova** [comment]
Closing stale PRs. Feel free to re-open if decide to continue working on this.

**Pull request closed without merging**